yt/frontends/gadget/fields.py

```diff
@@ -40,32 +46,121 @@ def _setup_four_metal_fractions(self, ptype):
             return data[(ptype, "FourMetalFractions")][:, i]
 
-        def _Metal_density(field, data, i: int):
+        def _metal_density(field, data, i: int):
             return data[(ptype, "FourMetalFractions")][:, i] * data[(ptype, "density")]
 
         for i, metal_name in enumerate(metal_names):
             # add the metal fraction fields
             self.add_field(
                 (ptype, metal_name + "_fraction"),
                 sampling_type="particle",
-                function=partial(_Fraction, i=i),
+                function=partial(_fraction, i=i),
                 units="",
             )
 
             # add the metal density fields
             self.add_field(
                 (ptype, metal_name + "_density"),
                 sampling_type="particle",
-                function=partial(_Metal_density, i=i),
+                function=partial(_metal_density, i=i),
                 units=self.ds.unit_system["density"],
             )
 
         return metal_names
 
-    def _setup_eleven_metal_masses(self, ptype):
+    def _make_fraction_functions(self, ptype, fname):
+        if fname == "ElevenMetalMasses":
+
+            def _fraction(field, data, i: int):
+                return (
+                    data[(ptype, "ElevenMetalMasses")][:, i]
+                    / data[(ptype, "particle_mass")]
+                )
+
+            def _metallicity(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")][:, 1].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return ret
+
+            def _h_fraction(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return 1.0 - ret
+
+            elem_names = ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"]
+
+        elif fname == "MetalMasses":
+            n_elem = len(
+                [
+                    fd
+                    for fd in self.ds.field_list
+                    if fd[0] == ptype and fd[1].startswith("MetalMasses")
+                ]
+            )
+            elem_names = {
+                7: ["C", "N", "O", "Mg", "Si", "Fe", "Ej"],
+                8: ["He", "C", "O", "Mg", "S", "Si", "Fe", "Ej"],
+                11: ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"],
+                15: [
+                    "He",
+                    "C",
+                    "Ca",
+                    "O",
+                    "N",
+                    "Ne",
+                    "Mg",
+                    "S",
+                    "Si",
+                    "Fe",
+                    "Na",
+                    "Al",
+                    "Ar",
+                    "Ni",
+                    "Ej",
+                ],
+            }[n_elem]
+            no_He = "He" not in elem_names
```

Comment:
I found the logic of this part a little bit confusing. But I don't have a good alternative suggestion.

Reply:
I agree with you--I can add some comments to explain what's going on.

Reply:
The new comments are helpful! Thank you for adding them.